### ISSUE TITLE
JupyterLab SoS kernel missing right panel kernel selections

### ISSUE THREAD

**Issue by gaow** [vatlab/jupyterlab-sos#72]
@BoPeng when you install latest components of SoS suite with Python 3.12 and JupyterLab 4.0, do you have issues with JupyterLab? This is what i experience: 

![image](https://github.com/vatlab/jupyterlab-sos/assets/917985/66c760ec-cb40-4cc7-aab1-45c6234d5bcd)

However my JupyterLab command output looks normal,

```
    To access the server, open this file in a browser:
        file:///home/gw/.local/share/jupyter/runtime/jpserver-1129777-open.html
    Or copy and paste one of these URLs:
        http://localhost:8890/lab?token=942df340c7ebdec4d45c4241c705aad5b7ced19a9e703849
        http://127.0.0.1:8890/lab?token=942df340c7ebdec4d45c4241c705aad5b7ced19a9e703849
[I 2024-04-11 20:12:46.743 ServerApp] Skipped non-installed server(s): bash-language-server, dockerfile-language-server-nodejs, javascript-typescript-langserver, julia-language-server, pyright, python-language-server, sql-language-server, texlab, typescript-language-server, unified-language-server, vscode-css-languageserver-bin, vscode-html-languageserver-bin, vscode-json-languageserver-bin, yaml-language-server
[W 2024-04-11 20:12:48.409 LabApp] Could not determine jupyterlab build status without nodejs
[I 2024-04-11 20:12:49.935 ServerApp] Kernel started: 1654bafd-95b6-4a3a-aefe-b646cd9a5d4f
/home/gw/micromamba/envs/python_libs/lib/python3.12/site-packages/sos_notebook/kernel.py:20: DeprecationWarning: pkg_resources is deprecated as an API. See https://setuptools.pypa.io/en/latest/pkg_resources.html
  import pkg_resources
0.00s - Debugger warning: It seems that frozen modules are being used, which may
0.00s - make the debugger miss breakpoints. Please pass -Xfrozen_modules=off
0.00s - to python to disable frozen modules.
0.00s - Note: Debugging will proceed. Set PYDEVD_DISABLE_FILE_VALIDATION=1 to disable this validation.
[I 2024-04-11 20:12:51.309 ServerApp] Connecting to kernel 1654bafd-95b6-4a3a-aefe-b646cd9a5d4f.
[I 2024-04-11 20:12:51.325 ServerApp] Connecting to kernel 1654bafd-95b6-4a3a-aefe-b646cd9a5d4f.
[I 2024-04-11 20:12:51.346 ServerApp] Connecting to kernel 1654bafd-95b6-4a3a-aefe-b646cd9a5d4f.
[I 2024-04-11 20:13:06.202 ServerApp] Creating new notebook in /code/mnm_analysis
[I 2024-04-11 20:13:06.599 ServerApp] Kernel started: 2989ac07-7386-4abf-ad6f-77a637ecbf7a
/home/gw/micromamba/envs/python_libs/lib/python3.12/site-packages/sos_notebook/kernel.py:20: DeprecationWarning: pkg_resources is deprecated as an API. See https://setuptools.pypa.io/en/latest/pkg_resources.html
  import pkg_resources
0.00s - Debugger warning: It seems that frozen modules are being used, which may
0.00s - make the debugger miss breakpoints. Please pass -Xfrozen_modules=off
0.00s - to python to disable frozen modules.
0.00s - Note: Debugging will proceed. Set PYDEVD_DISABLE_FILE_VALIDATION=1 to disable this validation.
[I 2024-04-11 20:13:07.920 ServerApp] Connecting to kernel 2989ac07-7386-4abf-ad6f-77a637ecbf7a.


[I 2024-04-11 20:14:53.367 ServerApp] Saving file at /code/mnm_analysis/Untitled.ipynb



[I 2024-04-11 20:16:53.437 ServerApp] Saving file at /code/mnm_analysis/Untitled.ipynb
[I 2024-04-11 20:18:53.523 ServerApp] Saving file at /code/mnm_analysis/Untitled.ipynb
```

Since for long time I've been running SoS notebook related components under JupyterLab 3.6 + Python 3.9, my recent upgrade seem to have various issues. I wonder if these issues can be reproduced on your end or it's just because I'm using some conda packages that are outdated even though my sos-notebook and jupyterlab-sos are the most up-to-date.

**Comment by gaow** [comment]
Looks like it's still an issue with my installation which uses older conda packages. For exmaple, SoS notebook cannot be upgrade on conda-forge because  ipykernel no longer ships a tests folder. So import ipykernel.tests throws an error.

Will use more specific tickets to follow up then revisit this when all is fixed.

**Comment by gaow** [comment]
Reopen this because even after i upgraded to latest jupyter-sos and sos-notebook it still does not work. No error from CLI terminal but there are no cell kernel choices and no syntax highlight.

**Comment by gaow** [comment]
Here are my versions relevant to SoS. 

```
  jupyterlab_sos                     0.10.1          py_0                 dnachun    
  sos                                0.25.1          pyhd8ed1ab_0         conda-forge
  sos-bash                           0.20.0          py_0                 conda-forge
  sos-notebook                       0.24.4          py_0                 dnachun    
  sos-python                         0.18.4          py_0                 conda-forge
  sos-r                              0.20.1          pyhd8ed1ab_0         conda-forge
  python                             3.12.2          hab00c5b_0_cpython   conda-forge
```

**Comment by BoPeng** [comment]
On my end,

* Python 3.12.0
* Jupyterlab 4.1.6
* jupyterlab-sos 0.10.1
* sos 0.25.1
* sos-notebook 0.24.4
* sos-pbs 0.20.8

You need to check 

```
jupyter kernelspec list --json
```
and see if the sos kernel is installed. Otherwise you need to run

```
python -m sos_notebook.install
```

Somehow I am seeing `jupyterlab-sos 0.9.1`

```
(py12) ➜  ~ jupyter labextension list
JupyterLab v4.1.6
/Users/bpeng/anaconda3/envs/py12/share/jupyter/labextensions
        jupyterlab-sos v0.9.0 enabled OK (python, jupyterlab_sos)
        transient-display-data v0.4.4 enabled OK (python, transient-display-data)
        jupyterlab_pygments v0.3.0 enabled OK (python, jupyterlab_pygments)
        @jupyter-notebook/lab-extension v7.1.2 enabled OK
        @jupyter-widgets/jupyterlab-manager v5.0.10 enabled OK (python, jupyterlab_widgets)
```

When you start a sos notebook, do you see any error message from console window (chrome, shift-right click, inspect, console)
![image](https://github.com/vatlab/jupyterlab-sos/assets/9889312/a8feb7e7-dbdc-440a-a296-16490ad45ce5)

**Comment by gaow** [comment]
Thanks @BoPeng for the detailed guide! To clarify, all packages i installed are from conda, including these packages temporarily, until we send in a PR: 

https://anaconda.org/dnachun/sos-notebook
https://anaconda.org/dnachun/jupyterlab-sos

I don't see in my labextension `jupyterlab-sos` ,

```
$ jupyter labextension list
JupyterLab v4.1.6
/home/gw/micromamba/envs/python_libs/share/jupyter/labextensions
        nbdime-jupyterlab v3.0.1 enabled OK
        jupyterlab_pygments v0.3.0 enabled OK (python, jupyterlab_pygments)
        transient-display-data v0.4.4 enabled OK (python, transient-display-data)
        jupyter-matplotlib v0.11.3 enabled OK
        ipyparallel-labextension v8.8.0 enabled OK
        @jupyter-widgets/jupyterlab-manager v5.0.10 enabled OK (python, jupyterlab_widgets)
        @jupyter-lsp/jupyterlab-lsp v5.1.0 enabled OK (python, jupyterlab-lsp)


Disabled extensions:
    @jupyterlab/completer-extension:base-service
    @jupyterlab/fileeditor-extension:language-server
    @jupyterlab/lsp-extension:settings
    @jupyterlab/notebook-extension:language-server
```
But in your list it was not the correct version either right? 

My kernels appear fine,

```
$ jupyter kernelspec list
Available kernels:
  bash            /home/gw/micromamba/envs/python_libs/share/jupyter/kernels/bash
  calysto_bash    /home/gw/micromamba/envs/python_libs/share/jupyter/kernels/calysto_bash
  ir              /home/gw/micromamba/envs/python_libs/share/jupyter/kernels/ir
  markdown        /home/gw/micromamba/envs/python_libs/share/jupyter/kernels/markdown
  python3         /home/gw/micromamba/envs/python_libs/share/jupyter/kernels/python3
  sos             /home/gw/micromamba/envs/python_libs/share/jupyter/kernels/sos
```

My console window looks like this 

![image](https://github.com/vatlab/jupyterlab-sos/assets/917985/9b0db2e9-84d7-4ad9-bae7-47d93a55d8a3)

Finally, to double check,

```
$ jupyter labextension list
JupyterLab v4.1.6
/home/gw/micromamba/envs/python_libs/share/jupyter/labextensions
        nbdime-jupyterlab v3.0.1 enabled OK
        jupyterlab_pygments v0.3.0 enabled OK (python, jupyterlab_pygments)
        transient-display-data v0.4.4 enabled OK (python, transient-display-data)
        jupyter-matplotlib v0.11.3 enabled OK
        ipyparallel-labextension v8.8.0 enabled OK
        @jupyter-widgets/jupyterlab-manager v5.0.10 enabled OK (python, jupyterlab_widgets)
        @jupyter-lsp/jupyterlab-lsp v5.1.0 enabled OK (python, jupyterlab-lsp)


Disabled extensions:
    @jupyterlab/completer-extension:base-service
    @jupyterlab/fileeditor-extension:language-server
    @jupyterlab/lsp-extension:settings
    @jupyterlab/notebook-extension:language-server
```

**Comment by gaow** [comment]
Is it what you see in the labextension folder? 

```
$ ls /home/gw/micromamba/envs/python_libs/share/jupyter/labextensions/jupyterlab-sos/
install.json
```
and
```
$ cat /home/gw/micromamba/envs/python_libs/share/jupyter/labextensions/transient-display-data/install.json 
{
  "packageManager": "python",
  "packageName": "transient-display-data",
  "uninstallInstructions": "Use your Python package manager (pip, conda, etc.) to uninstall the package transient-display-data"
}
```

It does not seem to have aactual contents. So maybe the installation did not work in the first place.

**Comment by gaow** [comment]
Update: if i force reinstall everything via `pip` including dependencies,

```
$ pip install jupyterlab-sos   --force-reinstall
```

Then everything indeed seems to work,

![image](https://github.com/vatlab/jupyterlab-sos/assets/917985/8e2d34a5-e6e3-450c-a681-d2b1f43b0f33)


although i still have the type error

![image](https://github.com/vatlab/jupyterlab-sos/assets/917985/379d1c88-7f12-4931-8732-0dd548e44698)

So, it seems that @BoPeng at least this type error, and the `jupyterlab-sos 0.9.0` confusions are the only issues using pip install. 

On our end, @denialnachun it seems the conda packages still did not work properly. I wonder why that is the case ... perhaps let's first PR what we have and see them together? The current conda packages on conda-forge are broken anyways so we are not making things any worse ...

**Comment by BoPeng** [comment]
Let me try again using conda.

**Comment by gaow** [comment]
Even for the pip version, i noticed that the previous highlight of _input _output etc is now gone,

<img width="813" alt="image" src="https://github.com/vatlab/jupyterlab-sos/assets/917985/df87df01-7695-4bf4-956f-b9bcfd9e027e">

If it is not intential then it might have to do with some of the errors we saw in the console?

**Comment by BoPeng** [comment]
> Even for the pip version, i noticed that the previous highlight of _input _output etc is now gone,

This is a known issue. JLab switched to CodeMirror6, which required a complete rewrite of the syntax highlighting stuff.. 

https://github.com/vatlab/jupyterlab-sos/issues/65#issuecomment-1720700731

**Comment by danielnachun** [comment]
@BoPeng We have fixed all of the issues here with the Conda package with some small tweaks to the feedstock recipe:
- We need to manually copy the `labextension` files into the right folder.  This is just an extra line in the recipe.  In our particular setup we also have to add a line to our `jupyter_lab_config.py`, but this is specific to our setup and not something that needs any changes to `jupyterlab-sos` or the Conda package.
- There were some missing dependencies for building with `hatchling`.
- One of the subpackages was missing `noarch: python`, so it was needlessly building platform specific packages.

I will push a PR with all these fixes to https://github.com/conda-forge/jupyterlab-sos-feedstock.

**Comment by dianacornejo** [comment]
@BoPeng as this issue is not closed yet and the problem is the same let me post here.

I've installed jupyter and SoS using the pixi+ micromamba install set up with @gaow and @danielnachun. 

Everything seems to install correctly with no error or warning messages. 

**Problem:** I Cannot see the cell level kernels in the drop down menu when using SoS

This is how my install looks like

```
dmc2245@csglogin:~$ jupyter labextension list
JupyterLab v4.2.0
/home/dmc2245/micromamba/envs/python_libs/share/jupyter/labextensions
        ipyparallel-labextension v8.8.0 enabled OK
        jupyter-matplotlib v0.11.4 enabled OK
        jupyterlab-sos v0.9.0 enabled OK (python, jupyterlab_sos)
        jupyterlab_pygments v0.3.0 enabled OK (python, jupyterlab_pygments)
        nbdime-jupyterlab v3.0.1 enabled OK
        transient-display-data v0.4.4 enabled OK (python, transient-display-data)
        @jupyter-lsp/jupyterlab-lsp v5.1.0 enabled OK (python, jupyterlab-lsp)
        @jupyter-widgets/jupyterlab-manager v5.0.10 enabled OK (python, jupyterlab_widgets)
Disabled extensions:
    @jupyterlab/completer-extension:base-service
    @jupyterlab/fileeditor-extension:language-server
    @jupyterlab/lsp-extension:settings
    @jupyterlab/notebook-extension:language-server
```


```
dmc2245@csglogin:~$ ls ~/micromamba/envs/python_libs/share/jupyter/labextensions
ipyparallel-labextension  jupyterlab_pygments  jupyterlab-sos  @jupyter-lsp  jupyter-matplotlib  @jupyter-widgets  nbdime-jupyterlab  transient-display-data
```

```
dmc2245@csglogin:~$ jupyter kernelspec list
Available kernels:
  bash            /mnt/vast/hpc/homes/dmc2245/.local/share/jupyter/kernels/bash
  calysto_bash    /mnt/vast/hpc/homes/dmc2245/.local/share/jupyter/kernels/calysto_bash
  ir              /mnt/vast/hpc/homes/dmc2245/.local/share/jupyter/kernels/ir
  markdown        /mnt/vast/hpc/homes/dmc2245/.local/share/jupyter/kernels/markdown
  python3         /mnt/vast/hpc/homes/dmc2245/.local/share/jupyter/kernels/python3
  sos             /mnt/vast/hpc/homes/dmc2245/.local/share/jupyter/kernels/sos
```

```
dmc2245@csglogin:~$ cat ~/micromamba/envs/python_libs/share/jupyter/labextensions/transient-display-data/install.json
{
  "packageManager": "python",
  "packageName": "transient-display-data",
  "uninstallInstructions": "Use your Python package manager (pip, conda, etc.) to uninstall the package transient-display-data"
}
```

Then I tried to apply the same @gaow did to re-install jupyterlab-sos. But it complained with the error message below 


```
dmc2245@csglogin:~$ pip install jupyterlab-sos   --force-reinstall
Collecting jupyterlab-sos
  Cache entry deserialization failed, entry ignored
  Could not find a version that satisfies the requirement jupyterlab-sos (from versions: )
No matching distribution found for jupyterlab-sos
```

I'm not sure how to get the drop-down kernels is SoS so I would appreciate any help on this! 

Thank you

**Comment by gaow** [comment]
During the past few days, both @dianacornejo and @danielnachun tried everything including installing via `pip` and `conda` on jjupyter-sos but still we don't get the cell level kernel menu. There may be something fundamentally wrong with jupyter-sos in terms of compatibility with other packages? It is particularly confusing since we were able to get it work flawlessly less than a month ago!

@danielnachun @dianacornejo could you post here your package environment with versions , so @BoPeng can verify on their end? But @BoPeng Im guessing if you start from scratch installing all the latest version of packages you may run into the issue. 

For conda based installation from scratch, this is the document we follow from: 

https://wanggroup.org/orientation/jupyter-setup#purge-previous-installations-of-micromamba-or-miniconda

Thanks in advance for your help!

**Comment by BoPeng** [comment]
You guys can at least have a look at the official sos-notebook docker image and compare, right?

**Comment by gaow** [comment]
are you refering to this file? https://github.com/vatlab/sos/blob/master/development/docker-notebook/Dockerfile its from years ago. Not sure how relevant it is.

**Comment by BoPeng** [comment]
Yes, I am talking about [the official sos-notebook docker image](https://vatlab.github.io/sos-docs/running.html#content). I can tell it is using jupyterlab 3.6.1 though. I can modify the docker file and upgrade it to jupyterlab 4.0. Hold on.

**Comment by gaow** [comment]
Here is what does not work, 

```
╰─ micromamba -n python_libs list | grep -P "sos|lab"
warning  libmamba 'root_prefix' set with default value: /Users/gw2411/micromamba
  jupyterlab                         4.2.1           pyhd8ed1ab_0          conda-forge
  jupyterlab-lsp                     5.1.0           pyhd8ed1ab_2          conda-forge
  jupyterlab-sos                     0.10.1          hd8ed1ab_0            conda-forge
  jupyterlab-transient-display-data  0.4.4           pyh5bfe37b_0          conda-forge
  jupyterlab_pygments                0.3.0           pyhd8ed1ab_1          conda-forge
  jupyterlab_server                  2.27.2          pyhd8ed1ab_0          conda-forge
  jupyterlab_sos                     0.10.1          pyh80e38bb_0          conda-forge
  jupyterlab_widgets                 3.0.11          pyhd8ed1ab_0          conda-forge
  sos                                0.25.1          pyhd8ed1ab_0          conda-forge
  sos-bash                           0.18.1          py_0                  conda-forge
  sos-notebook                       0.24.4          py_0                  dnachun
  sos-python                         0.20.2          pyhd8ed1ab_0          conda-forge
  sos-r                              0.20.1          pyhd8ed1ab_0          conda-forge
```

**Comment by BoPeng** [comment]
Yes, I updated our jupyterlab to 4.2.1 and jupyterlab-sos stops working. This is not supposed to happen (no api change for minor releases) but I will have a look later.

**Comment by gaow** [comment]
Thanks, glad we identified the cause of the issue now. We can either freeze our version for now until later we have a fix, or just hold on for the fix hoping it is not a huge issue!

**Comment by BoPeng** [comment]
I have an abstract due today, but can have a look this evening. Hopefully this is just a minor issue.

**Comment by danielnachun** [comment]
I think the full changelog for Jupyter 4.2 should be here: https://github.com/jupyterlab/jupyterlab/compare/v4.2.0rc0...8a17ca61cf0943225ddc4506763228dd21c90dab.  There's definitely no major changes so it's kind of hard to say what really broke...

**Comment by Ichoran** [comment]
I had the same problem and, following the advice here, found that using Jupyter 4.1.6 fixed the problem for new notebooks in JupyterLab.

Unfortunately, when I load an existing notebook, the dropdown doesn't appear.  But, luckily, toggling simple mode makes it reappear.  So there is a workaround.

However, if I switch to a non-SoS kernel, the menu doesn't go away.  Furthermore, toggling simple mode also doesn't make the menu disappear: it just stays in place.  Also, if I load a notebook with a non-SoS kernel and switch to SoS, toggling simple mode doesn't help;  I have to save, exit, reload, then toggle.

Whether with a new notebook or a pre-loaded one, if I switch from Jupyter Lab to Notebook and back, the drop-down menu is gone (whether or not it's a SoS kernel)--which kind of makes sense, since the switch is basically like loading a pre-existing file.  (Notebook never shows the menu for me--and it's 7.1.3, not 7.2.1.)

So something is a bit weird about how the menu customization is applied, even in JupyterLab 4.1.6.  It might be that 4.2 didn't actually introduce a bug, just changed things enough to reveal one that otherwise wouldn't always happen.

I wonder if there could be a race condition involved?  My computer is pretty fast (i9-14900HX etc) so parallel computations might come out in a different order for me than for others.  Without looking at any code, it feels very much like the interface is mutable data that the SoS kernel adds to but isn't refreshed when switching kernels, but when loading a whole notebook either SoS acts too soon and the change gets overwritten by a default, or too late and whatever it is modifying isn't used to generate the actual interface elements.  The simple mode switch happens with everything loaded, however, so the menu appears.  But whatever it is doesn't seem linked to the kernel itself, but rather loading a notebook with that kernel.

I have the exact same config now as @gaow had when it worked in April, save that I have `sos-python` at 0.20.2, and I don't have `jupyterlab-lsp` at all.

**Comment by BoPeng** [comment]
The issue turns out to be an CSS issue caused by some CSS adjustment from Jupyterlab 4.2.1. I have released jupyterlab-sos 0.10.2. Please let me know if the new release works.

**Comment by Ichoran** [comment]
@BoPeng - This only half fixes and half breaks it for me.  The conda package wasn't up yet, so I used pip install.  Now 7.2.1 will load the menu on a new notebook!  But loading an old notebook leaves me without the menu.  Furthermore, toggling simple mode no longer recovers it.

So it's progress of a sort, and seems to me to fix the original issue, but does not seem to fix my related issue.

**Comment by BoPeng** [comment]
@Ichoran  Please check if 0.10.3 has fixed #74.

**Comment by Ichoran** [comment]
@BoPeng - That fix works for me!

**Comment by gaow** [comment]
@BoPeng it works! However, if we have a very long command like this,

```
plink --bfile gwas_example_final --logistic hide-covar --covar gwas_example_final_pca.eigenvec --covar-number 1-10 --out gwas_example_final --ci 0.95
```

then the entire cell level kernel and other menus will be lost. I can only bring them back if we can make it multiple lines. 

@dianacornejo at least it works for most of the case if you try it out -- fresh install might be better?
